design/ceph/object/store.md

```diff
@@ -176,3 +183,50 @@ spec:
   gateway:
     caBundleRef: #ldaps-cabundle
 ```
+
+### Virtual Host Style access for buckets
+
+The Ceph Object Gateway supports accessing buckets using [virtual host style](https://docs.aws.amazon.com/AmazonS3/latest/userguide/VirtualHosting.html) which allows accessing buckets using the bucket name as a subdomain in the endpoint. The user can configure this option manually like below:
+
+```sh
+# ceph config set client.rgw.mystore.a rgw_dns_name my-ingress.mydomain.com,rook-ceph-rgw-my-store.rook-ceph.svc
+```
+
+ The hostname should valid characters codified by RFC-1123, and Rook Operator can perform input validation using k8s apimachinery `IsDNS1123Subdomain()`. Multiple endpoints can be added to the list separated by comma. But when the support is added Rook operator need to pass this arguments to RGW daemon by the flag `--rgw-dns-name`. This is supported from Ceph Reef release(v18.0) onwards. This is different from `customEndpoints` which is used for configuring the object store to replicate and sync data amongst Multisite. For accessing the bucket point user need to configure wildcard dns in the cluster using [ingress loadbalancer](https://kubernetes.io/docs/concepts/services-networking/ingress/#hostname-wildcards) or in openshift cluster use [dns operator](https://docs.openshift.com/container-platform/4.9/networking/dns-operator.html). Same for TLS certificate, user need to configure the TLS certificate for the wildcard dns for the RGW endpoint. This option won't be enabled by default, user need to enable it by adding the `hosting` section in the `Gateway` settings:
+
+```yaml
+spec:
+  gateway:
+    hosting:
+      dnsNames:
+      - "rook-ceph-rgw-my-store.rook-ceph.svc"
+      - "my-ingress.mydomain.com"
+```
+
+The user need to add all the required endpoints like the default k8s service endpoint, otherwise access to the bucket will fail. When this feature is enabled the endpoint in OBCs and COSI Bucket Access will be updated to virtual host style. The config map of the OBC and BucketInfo will look like below:
```

Comment:
Is this actually true? The default service does not support wildcards, so how is this going to work? 

AFAIU from my Q&A with @cbodley and Kyle Bader, the RGW will still respond to requests from other domains, but those domains will not support virtual-host style bucket addressing.

Reply:
```
$ ceph config get client.rgw.mystore.a
WHO                   MASK  LEVEL     OPTION                   VALUE        RO
global                      basic     log_to_file              false          
global                      advanced  mon_allow_pool_delete    true           
global                      advanced  mon_allow_pool_size_one  true           
global                      advanced  mon_cluster_log_file                    
client.rgw.mystore.a        advanced  rgw_dns_name             rgw.example  *
$ curl http://minikube:30802
<?xml version="1.0" encoding="UTF-8"?><Error><Code>NoSuchBucket</Code><Message></Message><BucketName>minikube</BucketName><RequestId>tx0000017a48b1a88bf1f0f-006569d164-215fb-my-store</RequestId><HostId>215fb-my-store-my-store</HostId></Error>[jiffin@fedora ~]$  
[jiffin@fedora ~]$ curl http://192.168.49.2:30802
<?xml version="1.0" encoding="UTF-8"?><ListAllMyBucketsResult xmlns="http://s3.amazonaws.com/doc/2006-03-01/"><Owner><ID>anonymous</ID><DisplayName></DisplayName></Owner><Buckets></Buckets></ListAllMyBucketsResult>
[jiffin@fedora ~]$ curl http://rgw.example
<?xml version="1.0" encoding="UTF-8"?><ListAllMyBucketsResult xmlns="http://s3.amazonaws.com/doc/2006-03-01/"><Owner><ID>anonymous</ID><DisplayName></DisplayName></Owner><Buckets></Buckets></ListAllMyBucketsResult>[jiffin@fedora ~]
```
The IPs are working fine but only mentioned hostname listed in `rgw_dns_name` is working fine

Reply:
> From the bug, it seems like the functionality may depend on the rgw_resolve_cname setting right?

there's an `if (g_conf()->rgw_resolve_cname` block just above that handles that, but the [block](https://github.com/ceph/ceph/blob/122c6b2/src/rgw/rgw_rest.cc#L2140-L2156) i linked above applies when that resolve-cname logic isn't successful

whenever the list of hostnames (from rgw_dns_name/rgw_dns_s3website_name etc) is non-empty, any requests with a non-matching host header are treated as vhost-style, using the entire host as the bucket name

Reply:
I feel like I'm still getting conflicting information, but maybe something is lost.

@thotz, if you remove `rook-ceph-rgw-my-store.rook-ceph.svc` from `dnsNames`, can you still use path-style access using the service endpoint?

I don't see a specific answer to that in your replies.

Reply:
Yes, in the above example, I have only set `rgw.example` as rgw_dns_name. `rook-ceph-rgw-my-store.rook-ceph.svc` is not in the list for rgw_dns_names

Reply:
Okay. I think that behavior is more clear, and more ideal. But we probably want some kind of assurance from Ceph that the behavior will remain consistent.

@cbodley is there a way to update the comments in the file you linked to be more clear about that behavior and to note that it is intended? Maybe link to this discussion? 

It seems to me like the `rgw_resolve_cname` can (maybe) control whether to allow path-style for "unknown" hosts in the header.

Reply:
something isn't consistent here.. in a local vstart setup, i started radosgw with `--rgw-dns-name example.com` and was unable to use path-style access as `localhost`:
```
~/ceph/build $ curl http://localhost:8000
<?xml version="1.0" encoding="UTF-8"?><Error><Code>NoSuchBucket</Code><Message></Message><BucketName>localhost</BucketName><RequestId>tx000002ac365575957c0f4-0065735463-4142-default</RequestId><HostId>4142-default-default</HostId></Error>
```
the radosgw log shows that localhost was interpreted as a bucket name:
```
2023-12-08T12:37:39.142-0500 7f3d0a4ee6c0 20 req 3081417995887034612 0.000000000s final domain/bucket subdomain=localhost domain= in_hosted_domain=1 in_hosted_domain_s3website=0 s->info.domain= s->info.request_uri=/localhost/
...
2023-12-08T12:37:39.143-0500 7f3d08ceb6c0 10 req 3081417995887034612 0.001000032s s->object=<NULL> s->bucket=localhost
```

Reply:
Yeah, that was a typo from my end instead of using `client.rgw.my.store.a` I have used `client.rgw.mystore.a` so the option was not set properly on the server. After enabling the `rgw_dns_name` name option, only the domain set mentioned in the list was accessible for the RGW server. Other domain names were failing but IPs are working fine.

options set on the rgw server:
```bash
$ ceph config get client.rgw.my.store.a
WHO                    MASK  LEVEL     OPTION                      VALUE        RO
client.rgw.my.store.a        advanced  debug_rgw                   20/20          
global                       basic     log_to_file                 false          
global                       advanced  mon_allow_pool_delete       true           
global                       advanced  mon_allow_pool_size_one     true           
global                       advanced  mon_cluster_log_file                       
client.rgw.my.store.a        advanced  rgw_dns_name              rook-ceph-rgw-my-store.rook-ceph.svc,rgw.example * 
client.rgw.my.store.a        advanced  rgw_enable_usage_log        true           
client.rgw.my.store.a        advanced  rgw_log_nonexistent_bucket  true           
client.rgw.my.store.a        advanced  rgw_log_object_name_utc     true           
client.rgw.my.store.a        advanced  rgw_run_sync_thread         true           
client.rgw.my.store.a        advanced  rgw_zone                    my-store     * 
client.rgw.my.store.a        advanced  rgw_zonegroup               my-store     * 
```

```bash

bash-4.4$ curl http://rook-ceph-rgw-my-store.rook-ceph.svc:80  --> default rgw service
<?xml version="1.0" encoding="UTF-8"?><ListAllMyBucketsResult xmlns="http://s3.amazonaws.com/doc/2006-03-01/"><Owner><ID>anonymous</ID><DisplayName></DisplayName></Owner><Buckets></Buckets></ListAllMyBucketsResult>bash-4.4$ 
bash-4.4$ curl http://rgw.example  --> ingress service
<?xml version="1.0" encoding="UTF-8"?><ListAllMyBucketsResult xmlns="http://s3.amazonaws.com/doc/2006-03-01/"><Owner><ID>anonymous</ID><DisplayName></DisplayName></Owner><Buckets></Buckets></ListAllMyBucketsResult>bash-4.4$ 
bash-4.4$ 
bash-4.4$ curl http://minikube:30802 --> minikube host Nodeport
<?xml version="1.0" encoding="UTF-8"?><Error><Code>NoSuchBucket</Code><Message></Message><BucketName>minikube</BucketName><RequestId>tx000001e15844eb4ed0c58-006576f3a9-2b42c-my-store</RequestId><HostId>2b42c-my-store-my-store</HostId></Error>bash-4.4$ 
bash-4.4$ 
bash-4.4$ curl http://rook-ceph-rgw-my-store-external:80 --> rgw service exposed via Nodeport
<?xml version="1.0" encoding="UTF-8"?><Error><Code>NoSuchBucket</Code><Message></Message><BucketName>rook-ceph-rgw-my-store-external</BucketName><RequestId>tx00000c367e30625370aa8-006576f3d6-2b42c-my-store</RequestId><HostId>2b42c-my-store-my-store</HostId></Error>bash-4.4$ 
```

IPs
```bash
$ curl  http://10.110.235.14:80 --> rook-ceph-rgw-my-store.rook-ceph.svc
<?xml version="1.0" encoding="UTF-8"?><ListAllMyBucketsResult xmlns="http://s3.amazonaws.com/doc/2006-03-01/"><Owner><ID>anonymous</ID><DisplayName></DisplayName></Owner><Buckets></Buckets></ListAllMyBucketsResult>bash-4.4$ 
bash-4.4$ 
bash-4.4$ 
bash-4.4$ curl  http://10.100.253.62:80 --> rook-ceph-rgw-my-store-external
<?xml version="1.0" encoding="UTF-8"?><ListAllMyBucketsResult xmlns="http://s3.amazonaws.com/doc/2006-03-01/"><Owner><ID>anonymous</ID><DisplayName></DisplayName></Owner><Buckets></Buckets></ListAllMyBucketsResult>bash-4.4$ 
bash-4.4$ 
bash-4.4$ 
bash-4.4$ curl  http://192.168.49.2:30802 --> minikube host
<?xml version="1.0" encoding="UTF-8"?><ListAllMyBucketsResult xmlns="http://s3.amazonaws.com/doc/2006-03-01/"><Owner><ID>anonymous</ID><DisplayName></DisplayName></Owner><Buckets></Buckets></ListAllMyBucketsResult>bash-4.4$ 
bash-4.4$ 
```

Reply:
Okay. We may want to plan for a helper in Rook then to auto-add the service if it exists. @travisn what are your thoughts?

Reply:
customEndpoints has the externally-routable names for multisite sync, right? sync uses path-style access, so setting rgw_dns_name to something else would break that. sounds like we'd want to add the customEndpoint names too

Reply:
@thotz I think what Casey and I are trying to get at here is that Rook should plan to automatically add the service endpoint (when the object store is in internal mode) and add the hostname component of all `customEndpoints` set on the CephObjectZone when any `dnsNames` are set to help users out. 

I think it also makes sense to do list deduplication as well in case users take it upon themselves to add those.

Please add detailed plans for those points to the design.

Travis and I will be on PTO for the upcoming Christmas holiday, so there is ample time for that.